### PR TITLE
script to create reference vcfs

### PR DESCRIPTION
**Pull request by ashwini06**
Given a reference variant text file in a specific format; python script extracts information from COSMIC file and creates the reference VCF file.

### CODE REVIEW

File: utilities/create_VCF/create_VCF.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from datetime import date
+
+AF = {}
+bg_fl = open(sys.argv[1], 'r')
+for ln in iter(bg_fl):
+    if not re.search('^Gene', ln):
+        ln = ln.strip().split('\t')
+        cos_id = ln[4] + ":" + ln[0] + ':' + ln[2]
+        AF[cos_id] = ln[3]
+bg_fl.close()
```
Comment (hassanfa) [inline utilities/create_VCF/create_VCF.py:25]: So this is ready column 4. What happens if column 4 is not AF?

Could add a bit of comment what exactly is being done here?

File: utilities/create_VCF/create_VCF.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+"##INFO=<ID=COSMIC,Number=1,Type=String,Description=\"Genomic Mutation Identifier\">\n"
+"##INFO=<ID=LEGACY_ID,Number=1,Type=String,Description=\"Legacy Mutation Identifier\">\n"
+"##INFO=<ID=CDS,Number=1,Type=String,Description=\"CDS mutation\">\n"
+"#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\n"
+)
+
+
+cosmic_lines = []
+with gzip.open('/home/proj/long-term-stage/cancer/UMI_analysis_runs_APJ/SeraCare_VCFs/cosmic_coding_muts_v89.vcf.gz', 'rt') as cosmic_fl:
+    for ln in iter(cosmic_fl):
```
Comment (hassanfa) [inline utilities/create_VCF/create_VCF.py:49]: This is going to iterate over all the variants in COSMIC. It'll take a while no? I suggest to use cyvcf2 or a similar package to make sure correct variant is found.

This part of script can be reused later to find out which variants are called in our sequenced sample.

File: utilities/create_VCF/create_VCF.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        ln = ln.strip()
+        if not re.search('^#',ln):
+            ln = ln.split('\t')
+            ginfo = ln[7].split(';')
+            gene = re.sub(r'(.*)(_ENST.*)',r'\1', ginfo[0])
+            gene2 = gene.replace('GENE=','')
+            aa = ginfo[4].replace('AA=','')
+            cos_id = ln[2] + ':' + gene2 + ':' + aa
```
Comment (hassanfa) [inline utilities/create_VCF/create_VCF.py:57]: packages like cyvcf2 can handle INFO field quite easily. Please have a look at that.

File: utilities/create_VCF/create_VCF.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            if cos_id in AF.keys():
+                af = "AF=" + str(round(float(AF[cos_id])/100,7)) 
+                info1 = ";".join([gene,ginfo[4],"VARIANT_TYPE=SNP",af,"AA_HGVS=","COSMIC="+ln[2],ginfo[2],ginfo[3]])
```
Comment (hassanfa) [inline utilities/create_VCF/create_VCF.py:60]: Empty field here can't be just empty, I think it is better to set it to ".".

**Comment by hassanfa** [comment]
@Mropat Could you also have a look and give a feedback on this?

**Comment by Mropat** [comment]
A few recommendations:

- As Hassan already mentioned, some packages (cyvcf2, pyvcf) provide support for reading, and iterating over VCF records, which can be useful for this task. 

- Suggest trying to contain each distinct task within their own function.
      - As the code-base grows larger, it becomes increasingly more risky to manipulate global variables in unintended ways, especially if you're preforming a lot of iterations. While this is not a problem currently, this future-proofs the code
      - Breaking up code in smaller chunks enables us to re-use the logic later if needed, without copying the code. Additionally, it will make it easier to write tests later.

- Look into Click library for easy and intuitive command line argument handling. Click along with argparse are widely used in many Clinical Genomics projects, and I personally was an instant convert after giving it a try.

**Comment by ashwini06** [comment]
@hassanfa @Mropat Thank you for the progressive comments and suggestions. I no wonder about the comments, as this code was initially written to achieve a particular task and thought its for a single time. cyvcf2 could minimize my script lines and I will make use of that. If we want to have a general implementation, I agree click+argsparse is a good option. I will work on this script further and update you once it's ready.

**Comment by hassanfa** [comment]
I know you had this script for SeraCare cases and thanks a lot for sharing so it can reused. This script does one thing and does it well. 👌 

I trust your judgment to see which parts of the comments can be achieved fairly quickly, and which parts of it do not need change (as it might have little value and take too much time). However, I'd appreciate if you could prioritize the following: 1. Structure the code a bit further (functions, comments, docstrings). 2. Add a main function. 3. A bit more descriptive variable names.

IMHO, use of Click, Argparse, cyvcf2, or pyvcf improvements are for maintainability and reproducibility reason.

**Comment by ashwini06** [comment]
Solved in https://github.com/Clinical-Genomics/BALSAMIC/pull/349

**Pull request closed without merging**